### PR TITLE
Sharapov Georgiy. Lab 4. Option 1.

### PR DESCRIPTION
**Pull request by nol0n**
Implemented mlir pass that breaks arith.ceildivui and arith.ceildivsi operations into calculation following the rule: ceildiv(a, b) = (a + b - 1) / b.

### CODE REVIEW

File: mlir/test/Transforms/lab4/sharapov_georgiy/test.mlir [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// RUN: mlir-opt -load-pass-plugin=%mlir_lib_dir/sCeilDiv%shlibext --pass-pipeline="builtin.module(sharapov_ceildiv)" %s | FileCheck %s
+
+module attributes {dlti.dl_spec = #dlti.dl_spec<#dlti.dl_entry<!llvm.ptr<270>, dense<32> : vector<4xi32>>, #dlti.dl_entry<i32, dense<32> : vector<2xi32>>, #dlti.dl_entry<f128, dense<128> : vector<2xi32>>, #dlti.dl_entry<!llvm.ptr<271>, dense<32> : vector<4xi32>>, #dlti.dl_entry<f64, dense<64> : vector<2xi32>>, #dlti.dl_entry<f16, dense<16> : vector<2xi32>>, #dlti.dl_entry<!llvm.ptr<272>, dense<64> : vector<4xi32>>, #dlti.dl_entry<i8, dense<8> : vector<2xi32>>, #dlti.dl_entry<i16, dense<16> : vector<2xi32>>, #dlti.dl_entry<i1, dense<8> : vector<2xi32>>, #dlti.dl_entry<!llvm.ptr, dense<64> : vector<4xi32>>, #dlti.dl_entry<i64, dense<64> : vector<2xi32>>, #dlti.dl_entry<f80, dense<128> : vector<2xi32>>, #dlti.dl_entry<"dlti.endianness", "little">, #dlti.dl_entry<"dlti.stack_alignment", 128 : i32>>} {
```
Comment (aobolensk) [inline mlir/test/Transforms/lab4/sharapov_georgiy/test.mlir]: These options may be omitted here.
```suggestion
module {
```

File: mlir/test/CMakeLists.txt [hunk 2]
```diff
@@ -86,7 +86,7 @@ configure_lit_site_cfg(
   ${CMAKE_CURRENT_SOURCE_DIR}/Unit/lit.cfg.py
   )
 
-set(MLIR_TEST_DEPENDS
+  set(MLIR_TEST_DEPENDS ${MLIR_TEST_DEPENDS}
```
Comment (aobolensk) [inline mlir/test/CMakeLists.txt]:
```suggestion
set(MLIR_TEST_DEPENDS ${MLIR_TEST_DEPENDS}
```

File: mlir/test/Transforms/lab4/sharapov_georgiy/test.mlir [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  llvm.func @ceildivsi(%arg0: i32, %arg1: i32) -> i32 attributes {llvm.linkage = #llvm.linkage<external>} {
+    // CHECK: llvm.func @ceildivsi(%arg0: i32, %arg1: i32) -> i32 attributes {llvm.linkage = #llvm.linkage<external>} {
+    // CHECK-NEXT: %c1_i32 = arith.constant 1 : i32
+    // CHECK-NEXT: %0 = arith.addi %arg0, %arg1 : i32
+    // CHECK-NEXT: %1 = arith.subi %0, %c1_i32 : i32
+    // CHECK-NEXT: %2 = arith.divsi %1, %arg1 : i32
```
Comment (aobolensk) [inline mlir/test/Transforms/lab4/sharapov_georgiy/test.mlir]: Do not rely on specific numbers in IR values:
```suggestion
    // CHECK-NEXT: %c1_i32 = arith.constant 1 : i32
    // CHECK-NEXT: %[[add:.*]] = arith.addi %arg0, %arg1 : i32
    // CHECK-NEXT: %[[sub:.*]] = arith.subi %[[add]], %c1_i32 : i32
    // CHECK-NEXT: %[[div:.*]] = arith.divsi %[[sub]], %arg1 : i32
```

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Review by aobolensk — APPROVED** [review]

File: mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    OpBuilder builder(op);
+    Location loc = op.getLoc();
+    Value a = op.getLhs();
+    Value b = op.getRhs();
+
+    Value one =
+        builder.create<arith::ConstantIntOp>(loc, 1, builder.getI32Type());
+    Value add = builder.create<arith::AddIOp>(loc, a, b);
+    Value sub = builder.create<arith::SubIOp>(loc, add, one);
+    Value div = builder.create<arith::DivUIOp>(loc, sub, b);
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp]: The difference between methods is only in one instruction. Would be nice to write one method with selections depending on input type to reduce code duplication.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    return "breaks arith.ceildivui and arith.ceildivsi operations into arith. "
+           "calculations";
+  }
+
+  void runOnOperation() override {
+    getOperation()->walk([&](Operation *op) {
+      if (auto ceilDivUI = dyn_cast<arith::CeilDivUIOp>(op)) {
+        replaceCeilDiv(ceilDivUI, OpType::UnsignedOp);
+      } else if (auto ceilDivSI = dyn_cast<arith::CeilDivSIOp>(op)) {
+        replaceCeilDiv(ceilDivSI, OpType::SignedOp);
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp:25]:
```suggestion
      if (isa<arith::CeilDivUIOp, arith::CeilDivSIOp>(op)) {
        replaceCeilDiv(dyn_cast<arith::IntBinaryOp>(op));
```

File: mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      }
+    });
+  }
+
+private:
+  template <typename CeilDivOpType>
+  void replaceCeilDiv(CeilDivOpType op, OpType type) {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp:32]:
```suggestion
  void replaceCeilDiv(arith::IntBinaryOp op) {
```

File: mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    Value a = op.getLhs();
+    Value b = op.getRhs();
+
+    Value one =
+        builder.create<arith::ConstantIntOp>(loc, 1, builder.getI32Type());
+    Value add = builder.create<arith::AddIOp>(loc, a, b);
+    Value sub = builder.create<arith::SubIOp>(loc, add, one);
+    Value div;
+
+    if (type == OpType::SignedOp) {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp:44]:
```suggestion
    if (isa<arith::CeilDivSIOp>(op)) {
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp (hunk 5 above)
Comment (nol0n) [reply mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp:25]: Initially, I wanted to write something similar. But I didn’t know if these operations were an implementation of some appropriate interface ([arith ceildiv ref](https://mlir.llvm.org/docs/Dialects/ArithOps/#arithceildivsi-arithceildivsiop)).

As far as I understand there are no such thing like BinaryOp or IntBinaryOp in arith dialect. And I can’t find anything similar that fits here.

```
error: no member named 'IntBinaryOp' in namespace 'mlir::arith'
                replaceCeilDiv(dyn_cast<arith::IntBinaryOp>(op));
                                        ~~~~~~~^
```

Comment (m-ly4) [reply mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp:25]: Ok, I see. Nevermind.

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**